usr/sbin/p11sak/p11sak.c

```diff
@@ -3481,27 +3481,181 @@ static const struct p11sak_objtype *find_certtype(CK_KEY_TYPE ktype)
     return NULL;
 }
 
-static CK_RV get_key_infos(CK_OBJECT_HANDLE key, CK_OBJECT_CLASS *class,
-                           CK_KEY_TYPE *ktype, CK_ULONG *keysize,
-                           char** label, char** typestr,
-                           const struct p11sak_objtype **keytype)
+static CK_RV get_common_name_value(CK_OBJECT_HANDLE obj, char *label,
+                                   char **common_name_value)
+{
+    CK_ATTRIBUTE attr = { CKA_VALUE, NULL, 0 };
+    X509 *x509 = NULL;
+    const CK_BYTE *tmp_ptr;
+    char *subj = NULL, *cn_tmp, *cn_tmp2 = NULL;
+    CK_RV rc;
+
+    rc = get_attribute(obj, &attr);
+    if (rc != CKR_OK) {
+        warnx("Failed to retrieve attribute CKA_VALUE from object "
+              "\"%s\": 0x%lX: %s", label, rc, p11_get_ckr(rc));
+        return rc;
+    }
+
+    if (attr.ulValueLen == 0 || attr.ulValueLen == CK_UNAVAILABLE_INFORMATION) {
+        warnx("CKA_VALUE of object \"%s\" is unavailable: 0x%lX: %s",
+              label, rc, p11_get_ckr(rc));
+        *common_name_value = strdup("");
+        if (*common_name_value == NULL) {
+            warnx("Failed to allocate memory for common_name_value "
+                  "for object \"%s\"", label);
+            return CKR_HOST_MEMORY;
+        }
+        return CKR_OK;
+    }
+
+    tmp_ptr = attr.pValue;
+    x509 = d2i_X509(NULL, &tmp_ptr, attr.ulValueLen);
+    if (x509 == NULL) {
+        *common_name_value = strdup("");
+        if (*common_name_value == NULL) {
+            warnx("Failed to allocate memory for common_name_value "
+                  "for object \"%s\"", label);
+            rc = CKR_HOST_MEMORY;
+        } else {
+            rc = CKR_OK;
+        }
+        goto done;
+    }
+
+    subj = X509_NAME_oneline(X509_get_subject_name(x509), NULL, 0);
+    *common_name_value = strdup(subj);
+    if (*common_name_value == NULL) {
+        warnx("Failed to allocate memory for common_name attribute"
+              "for object \"%s\"", label);
+        rc = CKR_HOST_MEMORY;
+        goto done;
+    }
+
+    cn_tmp = strstr(subj, "/CN=");
+    if (cn_tmp != NULL) {
+        cn_tmp2 = *common_name_value;
+        *common_name_value = strdup(cn_tmp + 4);
+        if (*common_name_value == NULL) {
+            warnx("Failed to allocate memory for common_name attribute"
+                  "for object \"%s\"", label);
+            rc = CKR_HOST_MEMORY;
+            goto done;
+        }
+    }
+
+    rc = CKR_OK;
+
+done:
+    free(attr.pValue);
+    if (subj != NULL)
+        OPENSSL_free(subj);
+    if (x509 != NULL)
+        X509_free(x509);
+    if (cn_tmp2 != NULL)
+        free(cn_tmp2);
+
+    return rc;
+}
+
+static CK_RV get_keysize_value(CK_OBJECT_HANDLE obj, char *label,
+                              const struct p11sak_objtype *objtype_val,
+                              CK_ULONG *keysize_val)
+{
+    CK_ATTRIBUTE keysize_attr;
+    CK_RV rc;
+
+    if (objtype_val->keysize_attr == (CK_ATTRIBUTE_TYPE) - 1) {
```

Comment:
Style: should be `(CK_ATTRIBUTE_TYPE)-1`, otherwise it looks like you subtract -1 from `(CK_ATTRIBUTE_TYPE)` :-)